### ISSUE TITLE
<locale>: Creating a module exporting header <locale> or <filesystem>does not work.

### ISSUE THREAD

**Issue by gmcode** [microsoft/STL#747]
Creating a module exporting header `<locale>` does not work. I'm assuming this should work?

If you try the same program bellow replacing `<locale>` with `<filesystem>` you will see other errors. It may be revealing to try all the headers.

Example of problem:
```
// mod.ixx
export module mod;
export {
#include <locale>
}
// mod.cpp
module mod;
```

To build:
```
rem works:
cl /nologo /c /experimental:module /std:c++latest /EHsc /MDd mod.ixx /Fomod.ifc.obj /module:output mod.ifc
rem fails:
cl /nologo /c /experimental:module /std:c++latest /EHsc /MDd mod.cpp /module:reference mod.ifc
```

Output:
```
mod.cpp(2): error C2535: 'void std::time_get<char,std::istreambuf_iterator<char,std::char_traits<char>>>::_Getvals<wchar_t>(wchar_t,
const std::_Locinfo &)': member function already defined or declared
C:\Program Files (x86)\Microsoft Visual Studio\2019\Preview\VC\Tools\MSVC\14.26.28801\include\xloctime(1029): note: see declaration
of 'std::time_get<char,std::istreambuf_iterator<char,std::char_traits<char>>>::_Getvals'
mod.cpp(2): error C2535: 'void std::time_get<wchar_t,std::istreambuf_iterator<wchar_t,std::char_traits<wchar_t>>>::_Getvals<wchar_t>
(wchar_t,const std::_Locinfo &)': member function already defined or declared
C:\Program Files (x86)\Microsoft Visual Studio\2019\Preview\VC\Tools\MSVC\14.26.28801\include\xloctime(1035): note: see declaration
of 'std::time_get<wchar_t,std::istreambuf_iterator<wchar_t,std::char_traits<wchar_t>>>::_Getvals'
```

Wondering:
Line 1029 of header `<xloctime>` has this code:
```
template _CRTIMP2_PURE void __CLR_OR_THIS_CALL time_get<char, istreambuf_iterator<char, char_traits<char>>>::_Getvals(
    wchar_t, const _Locinfo&);
    ^^^^^
```
Should this be char?

Other issue, similar circumstance, but not yet pinned down. I see errors like:

```
C:\Program Files (x86)\Microsoft Visual Studio\2019\Preview\VC\Tools\MSVC\14.26.28801\include\optional(132): error C3282: generic parameter lists can only appear on managed classes, structs, or functions
C:\Program Files (x86)\Microsoft Visual Studio\2019\Preview\VC\Tools\MSVC\14.26.28801\include\optional(150): note: see reference to class template instantiation 'std::_Optional_construct_base<_Ty>' being compiled
C:\Program Files (x86)\Microsoft Visual Studio\2019\Preview\VC\Tools\MSVC\14.26.28801\include\optional(141): error C3282: generic parameter lists can only appear on managed classes, structs, or functions
```

Finally: I also get this other error that seems module related, but I've not got a small reproducible case yet, but if this much error message is useful to you, please look into this:

```
C:\Program Files (x86)\Microsoft Visual Studio\2019\Preview\VC\Tools\MSVC\14.26.28801\include\optional(132): error C3282: generic parameter lists can only appear on managed classes, structs, or functions
C:\Program Files (x86)\Microsoft Visual Studio\2019\Preview\VC\Tools\MSVC\14.26.28801\include\optional(150): note: see reference to class template instantiation 'std::_Optional_construct_base<_Ty>' being compiled
C:\Program Files (x86)\Microsoft Visual Studio\2019\Preview\VC\Tools\MSVC\14.26.28801\include\optional(141): error C3282: generic parameter lists can only appear on managed classes, structs, or functions
myfile(67): error C2668: 'std::tie': ambiguous call to overloaded function
C:\Program Files (x86)\Microsoft Visual Studio\2019\Preview\VC\Tools\MSVC\14.26.28801\include\tuple(832): note: could be 
'std::tuple<const unsigned int &,const KeyCodeType &> std::tie<const unsigned int,const chessui::KeyCodeType>(const unsigned int &,const KeyCodeType &) noexcept'
C:\Program Files (x86)\Microsoft Visual Studio\2019\Preview\VC\Tools\MSVC\14.26.28801\include\tuple(832): note: or       'std::tuple<const unsigned int &,const KeyCodeType &> std::tie<const unsigned int,const KeyCodeType>(const unsigned int &,const KeyCodeType &) noexcept'
myfile(67): note: while trying to match the argument list '(const unsigned int, const KeyCodeType)'
```

I'm not using clr or 'generic' so I'm not sure why i'm seeing this. Sorry I can't give you more to go on yet with this message, but hopefully it's useful. The info after 'could be' is just some enum type, but in the error above both things it is talking about seem to be the same thing so i'm not sure what's ambiguous that its talking about?. Anyway, this part is all cryptic so I don't expect it's enough for you to go on, but it's all I have right now.

Thanks

**Comment by StephanTLavavej** [comment]
I investigated the `char/wchar_t` mismatch in that `time_get::_Getvals` export and filed #758. It's bogus but unused (exported for binary compatibility only), and isn't the cause of your module issues as far as I can tell (as the normal `wchar_t/wchar_t` signature is also mentioned). The trigger for that issue appears to be an explicit instantiation of a member function template of a class template.

These issues, especially the "error C3282: generic parameter lists can only appear on managed classes, structs, or functions", appear to be compiler issues with modules support. Please report them via [Developer Community](https://developercommunity.visualstudio.com/spaces/62/index.html) as explained in our README.

**Comment by gmcode** [comment]
Hi STL

Thanks for looking at this.

Would you mind forwarding the issue to someone on the module's team (over the partition lol) for me please, because otherwise it will remain broken, because I've looked at the issue reporting mechanism to report it myself but it still requests permissions to continue that I don't accept. I have in the past reported module's bugs directly before to GDR, but had no response and they've remained unfixed as far as what's been publicly released at least. I'm not prepared to expend further effort or use the bug reporting services until MS stop the permission insists just to report a bug; and I don't enjoy navigating MS's internal divisions like this when I do report bugs. MS should just be happy people take the time to report bugs with a solid repro case and notes rather than just make it harder for them them by subjecting them to forced agreements or navigating internal divisions to get bug reports noticed. Sorry.

By the way, I think if you use this test case for different headers I wouldn't be surprised if it revealed a few issues with a number of headers, but I've no idea how many might fall within your remit to fix.

**Comment by CaseyCarter** [comment]
FYI, this method of exporting the contents of a Standard Library header is forbidden by [[using.headers]/3](http://eel.is/c++draft/using.headers#3):

> A translation unit shall include a header only outside of any declaration or definition and, in the case of a module unit, only in its *global-module-fragment*, and shall include the header or import the corresponding header unit lexically before the first reference in that translation unit to any of the entities declared in that header. No diagnostic is required.

**Comment by gmcode** [comment]
Hi Casey
Thanks for this update. I queried 'I'm assuming this should work?' in my first message. fwiw I got the idea that it should work from here:
https://build2.org/build2/doc/build2-build-system-manual.xhtml#cxx-modules-intro
but re-reading it, I think I misunderstood what it was saying and it doesn't do this with standard headers only private ones.

Thank for clarifying the situation for me and especially for the Standard link.

Going forward, if anyone at MS is happy to receive direct emails regarding possible modules related bugs, do let me know. Sorry this module case was a red herring but at least one real non-module's bug came out of it.

Thanks again.
